src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.cs

```diff
@@ -12,37 +14,194 @@ internal static partial class ClrThreadPool
+
+                if(oldCounts == counts)
+                {
+                    return false;
+                }
+                counts = oldCounts;
+            }
+        }
+
+        private static void AdjustMaxWorkersActive()
+        {
+            int currentTicks = Environment.TickCount;
+            int totalNumCompletions = Volatile.Read(ref s_completionCount);
+            int numCompletions = totalNumCompletions - Volatile.Read(ref s_priorCompletionCount);
+            long startTime = s_currentSampleStartTime;
+            long endTime = 0; // TODO: PAL High Performance Counter
+            long freq = 0;
+
+            double elapsed = (double)(endTime - startTime) / freq;
+
+            if(elapsed * 1000 >= s_threadAdjustmentInterval / 2)
+            {
+                ThreadCounts currentCounts = ThreadCounts.VolatileReadCounts(ref s_counts);
+                int newMax;
+                (newMax, s_threadAdjustmentInterval) = HillClimbing.ThreadPoolHillClimber.Update(currentCounts.maxWorking, elapsed, numCompletions);
+
+                while(newMax != currentCounts.maxWorking)
+                {
+                    ThreadCounts newCounts = currentCounts;
+                    newCounts.maxWorking = (short)newMax;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_counts, newCounts, currentCounts);
+                    if (oldCounts == currentCounts)
+                    {
+                        if(newMax > oldCounts.maxWorking)
+                        {
+                            WorkerThread.MaybeAddWorkingWorker();
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        if(oldCounts.maxWorking > currentCounts.maxWorking && oldCounts.maxWorking >= newMax)
+                        {
+                            // someone (probably the gate thread) increased the thread count more than
+                            // we are about to do.  Don't interfere.
+                            break;
+                        }
+
+                        currentCounts = oldCounts;
+                    }
+                }
+                s_priorCompletionCount = totalNumCompletions;
+                Volatile.Write(ref s_nextCompletedWorkRequestsTime, currentTicks + s_threadAdjustmentInterval);
+                s_priorCompletedWorkRequestsTime = currentTicks;
+                s_currentSampleStartTime = endTime;
+            }
+        }
+
+        private static bool ShouldAdjustMaxWorkersActive()
+        {
+            // We need to subtract by prior time because Environment.TickCount can wrap around, making a comparison of absolute times unreliable.
+            int priorTime = s_priorCompletedWorkRequestsTime;
+            int requiredInterval = Volatile.Read(ref s_nextCompletedWorkRequestsTime) - priorTime;
```

Comment:
The volatile access here and in AdjustMaxWorkersActive should be on priorCompletedWorkRequestsTime instead. The main purpose here is to guarantee that the value read for the prior time in this function is never newer than the value read for the next time, thus implying that prior <= next. The other purpose of the full memory barrier used in the original code is to facilitate reading the latest value (volatile helps but does not guarantee that). I think volatile is fine for now, we can adjust it as needed.